arch/risc-v/src/mpfs/mpfs_corespi.c

```diff
@@ -0,0 +1,1544 @@
+static int mpfs_spi_trigger(struct spi_dev_s *dev)
+{
+  spierr("SPI trigger not supported\n");
+  DEBUGPANIC();
+
+  return -ENOSYS;
+}
+#endif
+
+/****************************************************************************
+ * Name: mpfs_spi_set_master_mode
+ *
+ * Description:
+ *   set spi mode
+ *
+ * Input Parameters:
+ *   master   - master or slave
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void mpfs_spi_set_master_mode(struct mpfs_spi_priv_s *priv,
+                                     uint8_t master)
+{
+  if (priv->enabled)
+    {
+      modifyreg32(MPFS_SPI_CONTROL, MPFS_SPI_ENABLE, 0);
+    }
+
+  if (master)
+    {
+      modifyreg32(MPFS_SPI_CONTROL, 0, MPFS_SPI_MODE);
+    }
+  else
+    {
+      modifyreg32(MPFS_SPI_CONTROL, MPFS_SPI_MODE, 0);
+    }
+
+  if (priv->enabled)
+    {
+      modifyreg32(MPFS_SPI_CONTROL, MPFS_SPI_ENABLE, MPFS_SPI_ENABLE);
+    }
+}
+
+/****************************************************************************
+ * Name: mpfs_spi_irq
+ *
+ * Description:
+ *   This is the common SPI interrupt handler. It will be invoked when an
+ *   interrupt is received on the device.
+ *
+ * Parameters:
+ *   cpuint        - CPU interrupt index
+ *   context       - Context data from the ISR
+ *   arg           - Opaque pointer to the internal driver state structure.
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success. A negated errno value is returned on
+ *   failure.
+ *
+ ****************************************************************************/
+
+static int mpfs_spi_irq(int cpuint, void *context, void *arg)
+{
+  struct mpfs_spi_priv_s *priv = (struct mpfs_spi_priv_s *)arg;
+  uint32_t remaining;
+  uint32_t status;
+  uint32_t rxretry;
+
+  status = getreg32(MPFS_SPI_INTMASK);
+
+  spiinfo("irq status=%x\n", status);
+
+  if (status & MPFS_SPI_DATA_RX)
+    {
+      remaining = priv->rxwords - priv->rx_pos;
+
+      if (remaining <= priv->fifosize)
+        {
+          mpfs_spi_unload_rx_fifo(priv, priv->rxbuf, remaining);
+        }
+      else
+        {
+          mpfs_spi_unload_rx_fifo(priv, priv->rxbuf, priv->fifolevel);
+        }
+
+      putreg32(MPFS_SPI_DATA_RX, MPFS_SPI_INT_CLEAR);
+    }
+
+  if (status & MPFS_SPI_TXDONE)
+    {
+      /* TX is done, we know RX is done too -> offload the RX FIFO */
+
+      rxretry = 0;
+
+try_rx_fifo_again:
+      remaining = priv->rxwords - priv->rx_pos;
+
+      if (remaining <= priv->fifosize)
+        {
+          mpfs_spi_unload_rx_fifo(priv, priv->rxbuf, remaining);
+        }
+      else
+        {
+          mpfs_spi_unload_rx_fifo(priv, priv->rxbuf, priv->fifolevel);
+        }
+
+      remaining = priv->rxwords - priv->rx_pos;
+      if (remaining > 0 && rxretry == 0)
+        {
+          /* Last RX character not ready yet, try again (once) */
+
+          rxretry = 1;
+          up_udelay(SPI_TTOA_US(priv->nbits + 2, priv->frequency));
```

Comment:
This is just hopelessly wrong. The granularity of PCLK_FREQ/DIVIDER is not near enough to use the _requested_ frequency here. The bias is too great. Must use the _actual_frequency.